src/vm_cpu.rs

```diff
@@ -314,8 +316,100 @@ impl VmPmuState {
             }
         }
     }
-}
 
+    /// Saves the internal state for PMU counters. Stops started counters, and resets all configured counters.
+    /// This should be called in anticipation of an outbound context switch.
+    fn save_counters(&mut self) {
+        use PmuCounterState::*;
+        fn reset_counter(counter_index: u64) -> SbiResult<()> {
+            sbi::api::pmu::stop_counters(
+                counter_index,
+                0x1,
+                PmuCounterStopFlags::default().set_reset_flag(),
+            )
+            .map_or_else(
+                |e| {
+                    // Treat already stopped error as success
+                    if matches!(e, SbiError::AlreadyStopped) {
+                        Ok(())
+                    } else {
+                        Err(e)
+                    }
+                },
+                |_| Ok(()),
+            )
```

Comment:
Do you actually need the `map` here? i.e. is `.or_else(|e| if matches!(e, SbiError::AlreadyStopped) { Ok(()) } else { Err(e) })` sufficient?